daemon/cmd/daemon_main.go

```diff
@@ -1563,24 +1568,97 @@ func initKubeProxyReplacementOptions() {
+// If it isn't and EnableAutoProtectNodePortRange == false, then return an error
+// making cilium-agent to stop.
+// Otherwise, if EnableAutoProtectNodePortRange == true, then append the nodeport
+// range to ip_local_reserved_ports.
 func checkNodePortAndEphemeralPortRanges() error {
-	val, err := sysctl.Read("net.ipv4.ip_local_port_range")
+	ephemeralPortRangeStr, err := sysctl.Read("net.ipv4.ip_local_port_range")
 	if err != nil {
 		return fmt.Errorf("Unable to read net.ipv4.ip_local_port_range")
 	}
-	ephemeralPortRange := strings.Split(val, "\t")
+	ephemeralPortRange := strings.Split(ephemeralPortRangeStr, "\t")
 	if len(ephemeralPortRange) != 2 {
-		return fmt.Errorf("Invalid ephemeral port range: %s", val)
+		return fmt.Errorf("Invalid ephemeral port range: %s", ephemeralPortRangeStr)
 	}
 	ephemeralPortMin, err := strconv.Atoi(ephemeralPortRange[0])
 	if err != nil {
 		return fmt.Errorf("Unable to parse min port value %s for ephemeral range", ephemeralPortRange[0])
 	}
-	if !(option.Config.NodePortMax < ephemeralPortMin) {
-		msg := `NodePort range (%d-%d) max port must be smaller than ephemeral range (%s-%s) min port. ` +
-			`Adjust ephemeral range port with "sysctl -w net.ipv4.ip_local_port_range='MIN MAX'".`
-		return fmt.Errorf(msg, option.Config.NodePortMin, option.Config.NodePortMax,
-			ephemeralPortRange[0], ephemeralPortRange[1])
+	ephemeralPortMax, err := strconv.Atoi(ephemeralPortRange[1])
+	if err != nil {
+		return fmt.Errorf("Unable to parse max port value %s for ephemeral range", ephemeralPortRange[1])
+	}
+
+	if option.Config.NodePortMax < ephemeralPortMin {
+		// ephemeral port range does not clash with nodeport range
+		return nil
 	}
+
+	nodePortRangeStr := fmt.Sprintf("%d-%d", option.Config.NodePortMin,
+		option.Config.NodePortMax)
+
+	if option.Config.NodePortMin > ephemeralPortMax {
+		return fmt.Errorf("NodePort port range (%s) is not allowed to be after ephemeral port range (%s)",
+			nodePortRangeStr, ephemeralPortRangeStr)
+	}
+
+	reservedPortsStr, err := sysctl.Read("net.ipv4.ip_local_reserved_ports")
+	if err != nil {
+		return fmt.Errorf("Unable to read net.ipv4.ip_local_reserved_ports")
+	}
+	for _, portRange := range strings.Split(reservedPortsStr, ",") {
+		if portRange == "" {
+			break
+		}
+		ports := strings.Split(portRange, "-")
+		if len(ports) == 0 {
+			return fmt.Errorf("Invalid reserved ports range")
+		}
+		from, err := strconv.Atoi(ports[0])
+		if err != nil {
+			return fmt.Errorf("Unable to parse reserved port %q", ports[0])
+		}
+		to := from
+		if len(ports) == 2 {
+			if to, err = strconv.Atoi(ports[1]); err != nil {
+				return fmt.Errorf("Unable to parse reserved port %q", ports[1])
+			}
+		}
+
+		if from <= option.Config.NodePortMin && to >= option.Config.NodePortMax {
+			// nodeport range is protected by reserved port range
+			return nil
+		}
+
+		if from > option.Config.NodePortMax {
+			break
+		}
+	}
```

Comment:
This can be a unit tests, you don't necessary need to write anything in `ip_local_reserved_ports` to unit test this.

Reply:
> This can be a unit tests,

Can you please elaborate more? Not sure whether I understood you.

Reply:
I checked with the `go test -cover` that I've covered all cases in this block (except for the obvious errors), so not sure whether it makes sense to move it to a separate function.

Reply:
I was not referring to the coverage. I was referring to the fact that `ip_local_reserved_ports` is being populated with random entries because of the unit testing.

Reply:
Populating the `ip_local_reserved_ports` from the tests should be harmless. Also, we restore previous values of it.